presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -208,41 +222,59 @@ else if (elementType == Slice.class) {
     }
 
     @UsedByGeneratedCode
-    public static Slice arrayJoin(MethodHandle castFunction, ConnectorSession session, Block arrayBlock, Slice delimiter)
+    public static Slice arrayJoin(
+            MethodHandle castFunction,
+            Object state,
+            ConnectorSession session,
+            Block arrayBlock,
+            Slice delimiter)
     {
-        return arrayJoin(castFunction, session, arrayBlock, delimiter, null);
+        return arrayJoin(castFunction, state, session, arrayBlock, delimiter, null);
     }
 
     @UsedByGeneratedCode
-    public static Slice arrayJoin(MethodHandle castFunction, ConnectorSession session, Block arrayBlock, Slice delimiter, Slice nullReplacement)
+    public static Slice arrayJoin(
+            MethodHandle castFunction,
+            Object state,
+            ConnectorSession session,
+            Block arrayBlock,
+            Slice delimiter,
+            Slice nullReplacement)
     {
+        PageBuilder pageBuilder = (PageBuilder) state;
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
         int numElements = arrayBlock.getPositionCount();
-
-        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(toIntExact(arrayBlock.getSizeInBytes() + delimiter.length() * arrayBlock.getPositionCount()));
+        BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(0);
 
         for (int i = 0; i < numElements; i++) {
             if (arrayBlock.isNull(i)) {
                 if (nullReplacement != null) {
-                    sliceOutput.appendBytes(nullReplacement);
+                    blockBuilder.writeBytes(nullReplacement, 0, nullReplacement.length());
                 }
                 else {
                     continue;
                 }
             }
             else {
                 try {
-                    sliceOutput.appendBytes((Slice) castFunction.invokeExact(arrayBlock, i, session));
+                    Slice slice = (Slice) castFunction.invokeExact(arrayBlock, i, session);
+                    blockBuilder.writeBytes(slice, 0, slice.length());
                 }
                 catch (Throwable throwable) {
+                    blockBuilder.closeEntry();
```

Comment:
I assume this is not necessary. It will be discarded anyway.

Reply:
@highker : When there is a TRY, we need to make sure the blockBuilder is closed correctly. See for example https://github.com/prestodb/presto/pull/8440 . I will add a comment :)